eZ/Publish/Core/Persistence/Cache/ObjectStateHandler.php

```diff
@@ -32,87 +34,97 @@ public function createGroup(InputStruct $input)
     {
-        $cacheItem = $this->cache->getItem('ez-state-group-' . $this->escapeForCacheKey($identifier) . '-by-identifier');
-        if ($cacheItem->isHit()) {
-            return $cacheItem->get();
-        }
-
-        $this->logger->logCall(__METHOD__, ['groupId' => $identifier]);
-        $group = $this->persistenceHandler->objectStateHandler()->loadGroupByIdentifier($identifier);
-
-        $cacheItem->set($group);
-        $cacheItem->tag(['state-group-' . $group->id]);
-        $this->cache->save($cacheItem);
-
-        return $group;
+        $escapedIdentifier = $this->escapeForCacheKey($identifier);
+
+        return $this->getCacheValue(
+            $identifier,
+            'ez-state-group-',
+            function (string $identifier): Group {
+                $this->logger->logCall(__METHOD__, ['groupId' => $identifier]);
+
+                return $this->persistenceHandler->objectStateHandler()->loadGroupByIdentifier($identifier);
+            },
+            static function (Group $group): array {
+                return ['state-group-' . $group->id];
+            },
+            static function (Group $group) use ($escapedIdentifier): array {
+                return ['ez-state-group-' . $escapedIdentifier . '-by-identifier', 'ez-state-group-' . $group->id];
+            },
+            '-by-identifier'
+        );
     }
 
     /**
      * {@inheritdoc}
      */
     public function loadAllGroups($offset = 0, $limit = -1)
     {
-        $cacheItem = $this->cache->getItem('ez-state-group-all');
-        if ($cacheItem->isHit()) {
-            return array_slice($cacheItem->get(), $offset, $limit > -1 ? $limit : null);
-        }
-
-        $this->logger->logCall(__METHOD__, ['offset' => $offset, 'limit' => $limit]);
-        $stateGroups = $this->persistenceHandler->objectStateHandler()->loadAllGroups(0, -1);
-
-        $cacheItem->set($stateGroups);
-        $cacheTags = [];
-        foreach ($stateGroups as $group) {
-            $cacheTags[] = 'state-group-' . $group->id;
-        }
-        $cacheItem->tag($cacheTags);
-        $this->cache->save($cacheItem);
-
-        return array_slice($stateGroups, $offset, $limit > -1 ? $limit : null);
+        $stateGroups = $this->getListCacheValue(
+            'ez-state-group-all',
+            function () use ($offset, $limit): array {
+                $this->logger->logCall(__METHOD__, ['offset' => (int) $offset, 'limit' => (int) $limit]);
+
+                return $this->persistenceHandler->objectStateHandler()->loadAllGroups(0, -1);
+            },
+            static function (Group $group): array {
+                return ['state-group-' . $group->id];
+            },
+            static function (Group $group): array {
+                return ['ez-state-group-' . $group->id, 'ez-state-group-' . $group->id . '-by-identifier'];
+            }
+        );
+
+        return \array_slice((array) $stateGroups, $offset, $limit > -1 ? $limit : null);
```

Comment:
Why the `(array)` type cast?

Reply:
https://github.com/ezsystems/ezpublish-kernel/pull/3106#discussion_r652617350

Reply:
It's impossible. @alongosz which static analysis tool reported that?

See `getListCacheValue` declaration:
```php
    /**
     * Load list of objects of some type and loggs the hits / misses.
     *
     * Load items from in-memory cache, symfony cache pool or backend in that order.
     * If not cached the returned objects will be placed in cache.
     *
     * @param string $key
     * @param callable $backendLoader Function for loading ALL objects, value is cached as-is.
     * @param callable $cacheTagger Gets cache object as argument, return array of cache tags.
     * @param callable $cacheIndexes Gets cache object as argument, return array of cache keys.
     * @param callable $listTags Optional, global tags for the list cache.
     * @param array $arguments Optional, arguments when parnt method takes arguments that key varies on.
     *
     * @return array
     */
    final protected function getListCacheValue(
        string $key,
        callable $backendLoader,
        callable $cacheTagger,
        callable $cacheIndexes,
        callable $listTags = null,
        array $arguments = []
    ) {
```

That's one thing. 

Second is that (assuming it can be an object) this type cast would in almost all cases invalid:

> If an object is converted to an array, the result is an array whose elements are the object's properties.
(see docs: https://www.php.net/manual/en/language.types.array.php#language.types.array.casting)

Reply:
@Steveb-p It must've been PHPStorm's internal one. I don't remember exactly, but when I looked over the code it seemed obvious. I need to restore previous state to see that.

Reply:
I agree with @Steveb-p here.

Reply:
> I agree with @Steveb-p here.

This is not a matter of agreeing or disagreeing. I think it requires a meeting so I can explain again why issues reported by IDE, and not PHPStan are also unacceptable.

Regarding the code itself, the state from the ee0c225fe6 commit, for which the comment was made, was:
```php
     $stateGroups = $this->getCacheValue(
            '',
            'ez-state-group-all',
            function () use ($offset, $limit): array {
                $this->logger->logCall(__METHOD__, ['offset' => (int) $offset, 'limit' => (int) $limit]);

                return $this->persistenceHandler->objectStateHandler()->loadAllGroups(0, -1);
            },
            static function (array $stateGroups): array {
                $cacheTags = [];
                foreach ($stateGroups as $group) {
                    $cacheTags[] = 'state-group-' . $group->id;
                }

                return $cacheTags;
            },
            static function (): array {
                return ['ez-state-group-all'];
            }
        );

        return \array_slice($stateGroups, $offset, $limit > -1 ? $limit : null);
```
I don't know at which point `getCacheValue` became `getListCacheValue`. Obviously for the current state even PHPStorm's analysis complains that `(array)` is redundant.

Reply:
Understood, then it seems we can safely remove the type cast. The misunderstanding came from changes in the method body, that subsequently changed the types.

```suggestion
        return \array_slice($stateGroups, $offset, $limit > -1 ? $limit : null);
```